pkg/runtime/pubsub/bulksub.go

```diff
@@ -0,0 +1,163 @@
+
+	defaultMaxBulkCount           int = 100
+	defaultMaxBulkAwaitDurationMs int = 1 * 1000
+)
+
+// msgWithCallback is a wrapper around a message that includes a callback function
+// that is called when the message is processed.
+type msgWithCallback struct {
+	msg contribPubsub.BulkMessageEntry
+	cb  func(error)
+}
+
+// defaultBulkSubscriber is the default implementation of BulkSubscriber.
+// It is used when the component does not implement BulkSubscriber.
+type defaultBulkSubscriber struct {
+	p contribPubsub.PubSub
+}
+
+// NewDefaultBulkSubscriber returns a new defaultBulkSubscriber from a PubSub.
+func NewDefaultBulkSubscriber(p contribPubsub.PubSub) *defaultBulkSubscriber {
+	return &defaultBulkSubscriber{
+		p: p,
+	}
+}
+
+// BulkSubscribe subscribes to a topic using a BulkHandler.
+// Dapr buffers messages in memory and calls the handler with a list of messages
+// when the buffer is full or max await duration is reached.
+func (p *defaultBulkSubscriber) BulkSubscribe(ctx context.Context, req contribPubsub.SubscribeRequest, handler contribPubsub.BulkHandler) error {
+	cfg := contribPubsub.BulkSubscribeConfig{
+		MaxBulkSubCount:                  utils.GetIntOrDefault(req.Metadata, maxBulkCountKey, defaultMaxBulkCount),
+		MaxBulkAwaitDurationMilliSeconds: utils.GetIntOrDefault(req.Metadata, maxBulkAwaitDurationMsKey, defaultMaxBulkAwaitDurationMs),
+	}
+
+	msgCbChan := make(chan msgWithCallback, cfg.MaxBulkSubCount)
+	go processBulkMessages(ctx, req.Topic, msgCbChan, cfg, handler)
+
+	// Subscribe to the topic and listen for messages.
+	return p.p.Subscribe(ctx, req, func(ctx context.Context, msg *contribPubsub.NewMessage) error {
+		entryID, err := uuid.NewRandom()
+		if err != nil {
+			return err
+		}
+
+		bulkMsg := contribPubsub.BulkMessageEntry{
+			EntryID:  entryID.String(),
+			Event:    msg.Data,
+			Metadata: msg.Metadata,
+		}
+
+		if msg.ContentType != nil {
+			bulkMsg.ContentType = *msg.ContentType
+		}
+
+		done := make(chan struct{})
+		msgCbChan <- msgWithCallback{
+			msg: bulkMsg,
+			cb: func(ierr error) {
+				err = ierr
+				close(done)
+			},
+		}
+
+		// Wait for the message to be processed.
+		<-done
+		return err
+	})
+}
+
+// processBulkMessages reads messages from msgChan and publishes them to a BulkHandler.
+// It buffers messages in memory and publishes them in bulk.
+func processBulkMessages(ctx context.Context, topic string, msgCbChan <-chan msgWithCallback, cfg contribPubsub.BulkSubscribeConfig, handler contribPubsub.BulkHandler) {
+	messages := make([]contribPubsub.BulkMessageEntry, cfg.MaxBulkSubCount)
+	msgCbMap := make(map[string]func(error), cfg.MaxBulkSubCount)
+
+	ticker := time.NewTicker(time.Duration(cfg.MaxBulkAwaitDurationMilliSeconds) * time.Millisecond)
+	defer ticker.Stop()
+
+	n := 0
+	for {
+		select {
+		case <-ctx.Done():
+			flushMessages(ctx, topic, messages[:n], msgCbMap, handler)
+			return
+		case msgCb := <-msgCbChan:
+			messages[n] = msgCb.msg
```

Comment:
I used indexing instead of `append` here because `n` is being used at multiple places here. It might be more readable (and as performant) as `append` in this case. WDYT"?

Reply:
Indexing is used instead of `append` as it also controls the scope of the slice being passed around. If we just change this one, it won't be any difference in perf. but will become slightly less readable, WDYT?

Reply:
Got it ... sorry did not notice `n=0`